drivers/filesystems/cdfs_new/shutdown.c

```diff
@@ -0,0 +1,175 @@
+    BOOLEAN VcbPresent;
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    //
+    //  Make sure we don't get any pop-ups.
+    //
+
+    SetFlag( IrpContext->Flags, IRP_CONTEXT_FLAG_DISABLE_POPUPS );
+
+    //
+    //  Initialize an event for doing calls down to
+    //  our target device objects.
+    //
+
+    KeInitializeEvent( &Event, NotificationEvent, FALSE );
+
+    //
+    //  Indicate that shutdown has started.
+    //
+
+    SetFlag( CdData.Flags, CD_FLAGS_SHUTDOWN );
+
+    //
+    //  Get everyone else out of the way
+    //
+
+    CdAcquireCdData( IrpContext );
+
+    //
+    //  Now walk through all the mounted Vcb's and shutdown the target
+    //  device objects.
+    //
+
+    Links = CdData.VcbQueue.Flink;
+
+    while (Links != &CdData.VcbQueue) {
+
+        Vcb = CONTAINING_RECORD( Links, VCB, VcbLinks );
+
+        //
+        //  Move to the next link now since the current Vcb may be deleted.
+        //
+
+        Links = Links->Flink;
+
+        //
+        //  If we have already been called before for this volume
+        //  (and yes this does happen), skip this volume as no writes
+        //  have been allowed since the first shutdown.
+        //
+
+        if (FlagOn( Vcb->VcbState, VCB_STATE_SHUTDOWN ) ||
+            (Vcb->VcbCondition != VcbMounted)) {
+
+            continue;
+        }
+
+        CdAcquireVcbExclusive( IrpContext, Vcb, FALSE );
+
+        CdPurgeVolume( IrpContext, Vcb, FALSE );
+
+        //
+        //  Build an irp for this volume stack - our own irp is probably too small and
+        //  each stack may have a different stack size.
+        //
+
+        NewIrp = IoBuildSynchronousFsdRequest( IRP_MJ_SHUTDOWN,
+                                               Vcb->TargetDeviceObject,
+                                               NULL,
+                                               0,
+                                               NULL,
+                                               &Event,
+                                               &Iosb );
+
+        if (NewIrp != NULL) {
+
+            Status = IoCallDriver( Vcb->TargetDeviceObject, NewIrp );
+
+            if (Status == STATUS_PENDING) {
+
+                (VOID)KeWaitForSingleObject( &Event,
+                                             Executive,
+                                             KernelMode,
+                                             FALSE,
+                                             NULL );
+            }
+
+            KeClearEvent( &Event );
+        }
+
+        SetFlag( Vcb->VcbState, VCB_STATE_SHUTDOWN );
+
+        //
+        //  Attempt to punch the volume down.
+        //
+
+        VcbPresent = CdCheckForDismount( IrpContext, Vcb, FALSE );
+
+        if (VcbPresent) {
+
+            CdReleaseVcb( IrpContext, Vcb );
+        }
+    }
+
+
+    CdReleaseCdData( IrpContext );
+
+    IoUnregisterFileSystem( CdData.FileSystemDeviceObject );
+    IoDeleteDevice( CdData.FileSystemDeviceObject );
```

Comment:
You also have to handle Hdd backed FS here.

Reply:
Hmm we didn't do that before in the old CdShutdown, but if it's needed...

Reply:
We did. Look at the commit history & the implementation.

Reply:
My bad, I re-read the old implementation, and I see that the logic was different, so yes, I missed something there. :P

Reply:
Note that you cannot mimic the exact same logic as previously as IoRegisterShutdownNotification() isn't used.